File: jupyter2/templates/asm/go/template.go
package how_compiled
import "math"

// (1)
func AddNums(x, y int64) int64 {
    return x + y + 123;
}

func AddInts(x, y int64) int64 {
    return x + y + 123;
}

func AddFloats(x, y float64) float64 {
    return x + y + 12.3
}

// (2)
type Point struct {
    x float64
    y float64
}

func GetPointY(p Point) float64 {
    return p.y
}

func GetPointYPtr(p * Point) float64 {
    return p.y
}

// (3)
func GetFloatArray3ElemConst(a [3]float64) float64 {
    return a[2]
}

func GetFloatArray3ElemVar(a [3]float64, i int64) float64 {
    return a[i]
}

func GetFloatSliceElemConst(a []float64) float64 {
    return a[2]
}

func GetFloatSliceElemVar(a []float64, i int64) float64 {
    return a[i]
}

// (4)
func Collatz(n int64) int64 {
	if n % 2 == 0 {
		return n / 2
	} else {
		return 3 * n + 1
	}
}

func Gcd1(a int64, b int64) int64 {
	if b == 0 {
		return a
	} else {
		return a % b
	}
}

// (5)
func SumArrayLoop(a []float64, n int64) float64 {
	s := 0.0
	var i int64;
	for i = 0; i < n; i++ {
		s += a[i]
	}
	return s;
}

// (6)
func CallTanh(x float64) float64 {
	return math.Tanh(x + 1.0) + x
}

// (7)
func SumArrayRec(a []float64, n int64) float64 {
	if (n == 0) {
		return 0.0
	} else {
		return SumArrayRec(a, n - 1) + a[n]
	}
}

// (8)
func SumArrayTail(a []float64, i int64, n int64, s float64) float64 {
	if i == n {
		return s
	} else {
		return SumArrayTail(a, i + 1, n, s + a[i])
	}
}

// (9)
func mkPoint(x, y float64) Point {
	p := Point{x: x + 1.0, y: y + 2.0}
	return p
}

func mkPointP(x, y float64) *Point {
	p := Point{x: x, y: y}
	return &p
}

// (10)
func mkArray10(x float64) [10]float64 {
	a := [10]float64{x, x, x, x, x, x, x, x, x, x}
	return a
}

func mkSlice10(x float64) []float64 {
	a := []float64{x, x, x, x, x, x, x, x, x, x}
	return a
}

// (11)
type Shape interface {
	area() float64
}

func CallArea(s Shape) float64 {
	return s.area()
}

func main() {
	println(CallTanh(4.0))
}

File: jupyter2/templates/asm/rs/template.rs
#![allow(dead_code)]

// (1)
#[no_mangle]
fn add_nums(x: i64, y: i64) -> i64 {
    x + y + 123
}

#[no_mangle]
fn add_ints(x: i64, y: i64) -> i64 {
    x + y + 123
}

#[no_mangle]
fn add_floats(x: f64, y: f64) -> f64 {
    x + y
}

// (2)

struct Point {
    x : f64,
    y : f64
}

#[no_mangle]
fn get_point_y(p : Point) -> f64 {
    p.y
}

#[no_mangle]
fn get_point_y_ref(p : &Point) -> f64 {
    p.y
}

#[no_mangle]
fn get_point_y_box(p : Box<Point>) -> f64 {
    p.y
}

// (3)

#[no_mangle]
fn get_float_array3_elem_const(a : &[f64; 3]) -> f64 {
    a[2]
}

#[no_mangle]
fn get_float_array3_elem_var(a : &[f64; 3], i : usize) -> f64 {
    a[i]
}

#[no_mangle]
fn get_float_slice_elem_const(a : &[f64]) -> f64 {
    a[2]
}

#[no_mangle]
fn get_float_slice_elem_var(a : &[f64], i : usize) -> f64 {
    a[i]
}

// (4)

#[no_mangle]
fn collatz(n : i64) -> i64 {
    if n % 2 == 0 {
	n / 2
    } else {
	3 * n + 1
    }
}

#[no_mangle]
fn gcd1(a : i64, b : i64) -> i64 {
    if b == 0 {
	a
    } else {
	a % b
    }
}

// (5)

#[no_mangle]
fn sum_array_loop(a : &[f64], n : usize) -> f64 {
    let mut s = 0.0;
    for i in 0..n {
	s += a[i];
    }
    s
}

// (6)

#[no_mangle]
fn call_tanh(x : f64) -> f64 {
    (x + 1.0).tanh() + x
}

// (7)

#[no_mangle]
fn sum_array_rec(a : &[f64], n : usize) -> f64 {
    if n == 0 {
	0.0
    } else {
	sum_array_rec(a, n - 1) + a[n]
    }
}

// (8)

#[no_mangle]
fn sum_array_tail(a : &[f64], i : usize, n : usize, s : f64) -> f64 {
    if i == n {
	s
    } else {
	sum_array_tail(a, i + 1, n, s + a[i])
    }
}

// (9)

#[no_mangle]
fn mk_point(x : f64, y : f64) -> Point {
    Point { x: x + 1.0, y: y + 2.0 }
}

#[no_mangle]
fn mk_point_box(x : f64, y : f64) -> Box<Point> {
    Box::new(Point { x: x + 1.0, y: y + 2.0 })
}

// (10)

#[no_mangle]
fn mk_array_10(x : f64) -> [f64; 10] {
    [x, x, x, x, x, x, x, x, x, x]
}

#[no_mangle]
fn mk_vec_10(x : f64) -> Vec<f64> {
    vec![x, x, x, x, x, x, x, x, x, x]
}

// (11)

trait Shape {
    fn area(&self) -> f64;
}

#[no_mangle]
fn call_area(s : &dyn Shape) -> f64 {
    s.area()
}

